webapp/src/mutator.ts

```diff
@@ -153,44 +153,51 @@ class Mutator {
-        //     async () => {
-        //         const res = await octoClient.insertBlock(block)
-        //         const jsonres = await res.json()
-        //         const newBlock = jsonres[0] as Block
-        //         await afterRedo?.(newBlock)
-        //         return newBlock
-        //     },
-    //     async (newBlock: Block) => {
-        //         await beforeUndo?.(newBlock)
-        //         await octoClient.deleteBlock(newBlock.id)
-        //     },
-    //     description,
-    //     this.undoGroupId,
-    // )
-        return createBoard()
-    }
+    async createBoardsAndBlocks(bab: BoardsAndBlocks, description = 'add', afterRedo?: (bab: BoardsAndBlocks) => Promise<void>, beforeUndo?: (bab: BoardsAndBlocks) => Promise<void>): Promise<BoardsAndBlocks> {
+        return undoManager.perform(
+            async () => {
+                const res = await octoClient.createBoardsAndBlocks(bab)
+                const newBab = (await res.json()) as BoardsAndBlocks
+                await afterRedo?.(newBab)
+                return newBab
+            },
+            async (newBab: BoardsAndBlocks) => {
+                await beforeUndo?.(newBab)
 
-    async updateBoard(boardId: string, newBoard: Board, oldBoard: Board, description: string): Promise<void> {
-        // ToDo: implement
+                const boardIds = newBab.boards.map(b => b.id)
+                const blockIds = newBab.blocks.map(b => b.id)
+                await octoClient.deleteBoardsAndBlocks(boardIds, blockIds)
+            },
+            description,
+            this.undoGroupId,
+        )
+    }
 
-        // const [updatePatch, undoPatch] = createPatchesFromBlocks(newBlock, oldBlock)
-        // await undoManager.perform(
-        //     async () => {
-        //         await octoClient.patchBlock(boardId, newBlock.id, updatePatch)
-        //     },
-    //     async () => {
-        //         await octoClient.patchBlock(boardId, oldBlock.id, undoPatch)
-        //     },
-    //     description,
-    //     this.undoGroupId,
-    // )
+    async updateBoard(newBoard: Board, oldBoard: Board, description: string): Promise<void> {
+        const [updatePatch, undoPatch] = createPatchesFromBoards(newBoard, oldBoard)
+        await undoManager.perform(
+            async () => {
+                await octoClient.patchBoard(newBoard.id, updatePatch)
+            },
+            async () => {
+                await octoClient.patchBoard(oldBoard.id, undoPatch)
+            },
+            description,
+            this.undoGroupId,
+        )
     }
 
     async deleteBoard(board: Board, description?: string, beforeRedo?: () => Promise<void>, afterUndo?: () => Promise<void>) {
-        // ToDo: implement
+        await undoManager.perform(
+            async () => {
+                await octoClient.deleteBoard(board.id)
+            },
+            async () => {
+                await octoClient.createBoard(board)
```

Comment:
I'm not 100% sure if this would work, because if you deleting a board, probably "undeleting" means creating it again with all the data again, or maybe we should have something like "undelete", that restore the last existing version from the history, or something like that, and the same for all the underneath children.

Reply:
I just based this change on the `deleteBlock` method, in which we're doing basically the same thing. Completely agree on the need of an "undelete", "unarchive" or whatever name we agree on type of method, but right now I think we can move forward with this solution as we're not changing the original behavior with it.